thriftbp/ttl_client.go

```diff
@@ -17,50 +22,135 @@ const DefaultMaxConnectionAgeJitter = 0.1
+	ttl       time.Duration
 
-	transport thrift.TTransport
+	replaceCounter metrics.Counter
 
-	// if expiration is zero, then the client will be kept open indefinetly.
-	expiration time.Time
+	// data guarded by lock (buffer-1 channel)
+	locked chan *ttlClientLockedData
 }
 
 // Close implements Client interface.
 //
 // It calls underlying TTransport's Close function.
 func (c *ttlClient) Close() error {
-	return c.transport.Close()
+	data := <-c.locked
+	defer func() {
+		c.locked <- data
+	}()
+	data.closed = true
+	if data.timer != nil {
+		data.timer.Stop()
+	}
+	return data.transport.Close()
+}
+
+func (c *ttlClient) Call(ctx context.Context, method string, args, result thrift.TStruct) (thrift.ResponseMeta, error) {
+	data := <-c.locked
+	defer func() {
+		c.locked <- data
+	}()
+	return data.client.Call(ctx, method, args, result)
 }
 
 // IsOpen implements Client interface.
 //
-// If TTL has passed, it closes the underlying TTransport and returns false.
-// Otherwise it just calls the underlying TTransport's IsOpen function.
+// It checks underlying TTransport's IsOpen first,
+// if that returns false, it returns false.
+// Otherwise it checks TTL,
+// returns false if TTL has passed and also close the underlying TTransport.
 func (c *ttlClient) IsOpen() bool {
-	if !c.transport.IsOpen() {
+	data := <-c.locked
+	defer func() {
+		c.locked <- data
+	}()
+	if !data.transport.IsOpen() {
 		return false
 	}
-	if !c.expiration.IsZero() && time.Now().After(c.expiration) {
-		c.transport.Close()
+	if !data.expiration.IsZero() && time.Now().After(data.expiration) {
+		data.transport.Close()
 		return false
 	}
 	return true
 }
 
-// newTTLClient creates a ttlClient with a thrift TTransport and a ttl+jitter.
-func newTTLClient(transport thrift.TTransport, client thrift.TClient, ttl time.Duration, jitter float64) *ttlClient {
-	var expiration time.Time
+// refresh is called when the ttl hits to try to refresh the connection.
+func (c *ttlClient) refresh() {
+	client, transport, err := c.generator()
+	if err != nil {
+		// We cannot replace this connection in the background,
+		// leave client and transport be,
+		// this connection will be replaced by the pool upon next use.
+		c.replaceCounter.With("success", "False").Add(1)
+		return
+	}
+
+	// replace with the refreshed connection
+	data := <-c.locked
+	defer func() {
+		c.locked <- data
+	}()
+	if data.closed {
+		// If Close was called after we entered this function,
+		// close the newly created connection and return early.
+		transport.Close()
+		return
+	}
+	c.renew(time.Now(), data)
+	data.client = client
+	if data.transport != nil {
+		// close the old transport before replacing it, to avoid connection leaks.
+		data.transport.Close()
+	}
+	data.transport = transport
+	c.replaceCounter.With("success", "True").Add(1)
+}
+
+// renew updates expiration and timer in data base on the given timestamp.
+func (c *ttlClient) renew(now time.Time, data *ttlClientLockedData) {
```

Comment:
I typically reverse this -- `*ttlClientLockedData` is the receiver, and then it can take in a *ttlClient param